src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java

```diff
@@ -61,12 +62,28 @@ public void cancel() { }
     final BodySubscriber<T> userSubscriber;
     final AtomicBoolean completed = new AtomicBoolean();
     final AtomicBoolean subscribed = new AtomicBoolean();
-    volatile Subscription subscription;
+    volatile SubscriptionWrapper subscription;
     volatile Throwable withError;
     public HttpBodySubscriberWrapper(BodySubscriber<T> userSubscriber) {
         this.userSubscriber = userSubscriber;
     }
 
+    class SubscriptionWrapper implements Subscription {
+        final Subscription subscription;
+        SubscriptionWrapper(Subscription s) {
+            this.subscription = Objects.requireNonNull(s);
+        }
+        @Override
+        public void request(long n) {
+            subscription.request(n);
+        }
+
+        @Override
+        public void cancel() {
+            HttpBodySubscriberWrapper.this.cancel(subscription);
```

Comment:
Hello Daniel, at first when I noticed this code, it looked odd to me that we aren't calling the underlying `subscription.cancel()`. I then noticed that the `HttpBodySubscriberWrapper.cancel()` actually does the cancellation of this underyling `subscription`. It's slightly odd to me that we are doing it there instead of here. Furthermore, the `HttpBodySubscriberWrapper` (which implements the `Flow` interface) has APIs like `onComplete()`, `onError()`. Perhaps it would be more consistent to have an `onCancel` in `HttpBodySubscriberWrapper` and then have the actual cancellation of the `subscription` done here in the `SubscriptionWrapper` and then call the `onCancel(Subscription)` on the `HttpBodySubscriberWrapper`? Something like:



```
class SubscriptionWrapper ... {
  @Override
  public void cancel() {
    try {
      subscription.cancel();
    } finally {
         HttpBodySubscriberWrapper.this.onCancel(subscription);
    }
  }
}
```

```
class HttpBodySubscriberWrapper {
    protected void onCancel(Subscription cancelledSubscription) {
        
    }
}

Reply:
good idea. done.